### PR TITLE
Make `String(::Memory)` copy

### PR DESCRIPTION
**Pull request by nhz2**
A more targeted fix of #54369 than #54372

Preserves the performance improvements added in #53962 by creating a new internal `unsafe_takestring(v::Memory{UInt8})` function that does what `String(::Memory{UInt8})` used to do.

### CODE REVIEW

**Comment by nhz2** [comment]
bump

**Comment by nhz2** [comment]
bump

**Comment by nhz2** [comment]
bump

**Comment by nhz2** [comment]
bump

**Comment by nhz2** [comment]
bump

File: stdlib/FileWatching/src/pidfile.jl [hunk 1]
```diff
@@ -280,7 +280,7 @@ function open_exclusive(path::String;
 end
 
 function _rand_filename(len::Int=4) # modified from Base.Libc
-    slug = Base.StringMemory(len)
+    slug = Base.StringVector(len)
```
Comment (jakobnissen) [inline stdlib/FileWatching/src/pidfile.jl:283]: Overall, looks good to me. I wonder why this here uses `StringVector` over `StringMemory`? As far as I can see, the memory can never escape the function.

**Review by jakobnissen — APPROVED** [review]
Everything looks good to me. Would much prefer to get this into 1.11

File: stdlib/FileWatching/src/pidfile.jl (hunk 1 above)
Comment (nhz2) [reply stdlib/FileWatching/src/pidfile.jl:283]: Since this is not performance-sensitive I wanted to avoid using the internal function here.

**Comment by nhz2** [comment]
@jakobnissen thank you for reviewing this. I'd also like to get this in 1.11

**Comment by DilumAluthge** [comment]
I think it would be ideal if we could get @vtjnash to review this before merging.

**Comment by DilumAluthge** [comment]
@vtjnash Can you review this?

**Comment by DilumAluthge** [comment]
Bump @vtjnash

**Comment by nhz2** [comment]
This is an important bug fix because the documentation in Julia 1.11 explicitly states that the `String` constructor will copy for all `AbstractVectors{UInt8}` except `Vector{UInt8}`.

**Comment by DilumAluthge** [comment]
Bump @vtjnash Can you review this?

**Pull request merged**